graylog2-server/src/main/java/org/graylog2/rest/resources/system/field_types/FieldTypeChangeResource.java

```diff
@@ -0,0 +1,147 @@
+    private final MongoIndexSet.Factory mongoIndexSetFactory;
+    private final MongoIndexSetService mongoIndexSetService;
+
+    @Inject
+    public FieldTypeChangeResource(final IndexSetService indexSetService,
+                                   final MongoIndexSet.Factory mongoIndexSetFactory,
+                                   final MongoIndexSetService mongoIndexSetService) {
+        this.indexSetService = indexSetService;
+        this.mongoIndexSetFactory = mongoIndexSetFactory;
+        this.mongoIndexSetService = mongoIndexSetService;
+    }
+
+    @PUT
+    @Path("{field_name}")
+    @Timed
+    @ApiOperation(value = "Change field type for certain index sets")
+    @ApiResponses(value = {
+            @ApiResponse(code = 403, message = "Unauthorized")
+    })
+    @NoAuditEvent("No audit for field type changes")
+    public Response changeFieldType(@ApiParam(name = "index_sets_ids", required = true)
+                                    @QueryParam("index_sets_ids") Set<String> indexSetsIds,
+                                    @ApiParam(name = "field_name", required = true)
+                                    @PathParam("field_name") String fieldName,
+                                    @ApiParam(name = "new_type", required = true)
+                                    @QueryParam("new_type") String newType,
+                                    @ApiParam(name = "rotate_immediately", defaultValue = "false")
+                                    @QueryParam("rotate_immediately") boolean rotateImmediately
+    ) {
+        checkPermissions(indexSetsIds);
+
+        //TODO: validation of fields
+
+        CustomFieldMappings.CustomFieldMapping customMapping = new CustomFieldMappings.CustomFieldMapping(fieldName, newType);
+
+        for (String indexSetId : indexSetsIds) {
+            try {
+                final Optional<IndexSetConfig> indexSetConfigOpt = indexSetService.get(indexSetId);
+                if (indexSetConfigOpt.isPresent()) {
+                    final IndexSetConfig indexSetConfig = indexSetConfigOpt.get();
+                    final IndexSetConfig updatedIndexSetConfig = storeMapping(customMapping, indexSetConfig);
+
+                    if (rotateImmediately) {
+                        cycleIndexSet(updatedIndexSetConfig);
+                    }
+                }
+            } catch (Exception ex) {
+                LOG.error("Failed to update field type in index set : " + indexSetId, ex);
+            }
+        }
+
+        return Response.ok().build();
+    }
+
+    private IndexSetConfig storeMapping(final CustomFieldMappings.CustomFieldMapping customMapping,
+                                        final IndexSetConfig indexSetConfig) {
+        final IndexSetConfig.Builder builder = indexSetConfig.toBuilder();
+        final CustomFieldMappings previousCustomFieldMappings = indexSetConfig.customFieldMappings();
+        if (previousCustomFieldMappings == null) {
```

Comment:
If we create an empty instance of `CustomFieldMappings` in the `IndexSetConfig` model when there is none present, we can avoid this check.